File: code/residentadvisor_api_examples.js
/**
 * Ejemplos de código para Resident Advisor API
 * Incluye ejemplos en JavaScript/Node.js y Bash
 */

// ============================================
// JAVAScript/NODE.js
// ============================================

const axios = require('axios');

class ResidentAdvisorAPI {
    constructor() {
        this.endpoint = 'https://ra.co/graphql';
        this.headers = {
            'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36',
            'Content-Type': 'application/json'
        };
    }

    /**
     * Obtiene eventos populares
     */
    async getPopularEvents(filters = {}, pageSize = 20) {
        const query = `
            query GET_POPULAR_EVENTS($filters: FilterInputDtoInput, $pageSize: Int!) {
                popularEvents(filters: $filters, pageSize: $pageSize) {
                    id
                    title
                    attending
                    date
                    contentUrl
                    flyerFront
                    images {
                        id
                        filename
                        alt
                        type
                        crop
                    }
                    venue {
                        id
                        name
                        contentUrl
                        live
                    }
                }
            }
        `;

        const variables = {
            filters,
            pageSize
        };

        try {
            const response = await axios.post(this.endpoint, {
                query,
                variables
            }, {
                headers: this.headers
            });

            return response.data;
        } catch (error) {
            console.error('Error al hacer request:', error.message);
            return null;
        }
    }

    /**
     * Obtiene eventos por área específica
     */
    async getEventsByArea(area, pageSize = 20) {
        const filters = { area };
        return this.getPopularEvents(filters, pageSize);
    }

    /**
     * Obtiene eventos por fecha de listado
     */
    async getEventsByListingDate(listingDate, pageSize = 20) {
        const filters = { listingDate };
        return this.getPopularEvents(filters, pageSize);
    }
}

// Ejemplo de uso
async function ejemploJavaScript() {
    console.log('=== EJEMPLO JAVASCRIPT/NODE.js ===');
    
    const raApi = new ResidentAdvisorAPI();
    
    try {
        // Obtener eventos populares
        const eventos = await raApi.getPopularEvents({}, 10);
        
        if (eventos && eventos.data && eventos.data.popularEvents) {
            console.log(`Se encontraron ${eventos.data.popularEvents.length} eventos:`);
            
            eventos.data.popularEvents.forEach((evento, index) => {
                console.log(`${index + 1}. ${evento.title}`);
                console.log(`   Fecha: ${evento.date}`);
                console.log(`   Asistentes: ${evento.attending}`);
                console.log(`   Lugar: ${evento.venue.name}`);
                console.log('');
            });
        }
        
        // Ejemplo con filtros
        console.log('=== EJEMPLO CON FILTROS ===');
        const eventosLondon = await raApi.getEventsByArea('London', 5);
        
        if (eventosLondon && eventosLondon.data) {
            console.log(`Eventos en London: ${eventosLondon.data.popularEvents?.length || 0}`);
        }
        
    } catch (error) {
        console.error('Error:', error.message);
    }
}

// ============================================
// BASH/CURL
// ============================================

const ejemplosBash = {
    /**
     * Ejemplo básico de CURL
     */
    comandoCurlBasico: `curl --location 'https://ra.co/graphql' \\
--header 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36' \\
--header 'Content-Type: application/json' \\
--data '{
  "query": "query GET_POPULAR_EVENTS($filters: FilterInputDtoInput, $pageSize: Int!) { popularEvents(filters: $filters, pageSize: $pageSize) { id title attending date contentUrl flyerFront images { id filename alt type crop } venue { id name contentUrl live } } }",
  "variables": {
    "filters": {},
    "pageSize": 20
  }
}'`,

    /**
     * CURL con filtros por área
     */
    comandoCurlConFiltros: `curl --location 'https://ra.co/graphql' \\
--header 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36' \\
--header 'Content-Type: application/json' \\
--data '{
  "query": "query GET_POPULAR_EVENTS($filters: FilterInputDtoInput, $pageSize: Int!) { popularEvents(filters: $filters, pageSize: $pageSize) { id title attending date contentUrl flyerFront images { id filename alt type crop } venue { id name contentUrl live } } }",
  "variables": {
    "filters": {"area": "London"},
    "pageSize": 10
  }
}'`,

    /**
     * CURL con formato legible
     */
    comandoCurlFormateado: `curl --location 'https://ra.co/graphql' \\
  --header 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36' \\
  --header 'Content-Type: application/json' \\
  --data '{
    "query": "query GET_POPULAR_EVENTS($filters: FilterInputDtoInput, $pageSize: Int!) { popularEvents(filters: $filters, pageSize: $pageSize) { id title attending date contentUrl flyerFront images { id filename alt type crop } venue { id name contentUrl live } } }",
    "variables": {
      "filters": {
        "area": "Berlin",
        "listingDate": "2024-12-01"
      },
      "pageSize": 15
    }
  }'`
};

/**
 * Función para mostrar todos los ejemplos bash
 */
function mostrarEjemplosBash() {
    console.log('=== EJEMPLOS BASH/CURL ===');
    console.log('');
    console.log('1. Comando básico:');
    console.log(ejemplosBash.comandoCurlBasico);
    console.log('');
    console.log('2. Con filtros por área:');
    console.log(ejemplosBash.comandoCurlConFiltros);
    console.log('');
    console.log('3. Con formato legible y múltiples filtros:');
    console.log(ejemplosBash.comandoCurlFormateado);
    console.log('');
}

// ============================================
// UTILIDADES
// ============================================

/**
 * Función para hacer requests con retry
 */
async function makeRequestWithRetry(api, query, variables, maxRetries = 3) {
    for (let i = 0; i < maxRetries; i++) {
        try {
            const response = await api.getPopularEvents(variables);
            return response;
        } catch (error) {
            console.log(`Intento ${i + 1} falló: ${error.message}`);
            if (i === maxRetries - 1) {
                throw error;
            }
            await new Promise(resolve => setTimeout(resolve, 1000 * (i + 1)));
        }
    }
}

/**
 * Función para procesar múltiples requests con delays
 */
async function processMultipleRequests(api, requests, delayMs = 1000) {
    const results = [];
    
    for (let i = 0; i < requests.length; i++) {
        const request = requests[i];
        console.log(`Procesando request ${i + 1}/${requests.length}`);
        
        try {
            const result = await api.getPopularEvents(request.filters, request.pageSize);
            results.push({
                request: request,
                result: result,
                success: true
            });
        } catch (error) {
            results.push({
                request: request,
                error: error.message,
                success: false
            });
        }
        
        // Delay entre requests para evitar rate limiting
        if (i < requests.length - 1) {
            await new Promise(resolve => setTimeout(resolve, delayMs));
        }
    }
    
    return results;
}

// ============================================
// EJECUCIÓN DE EJEMPLOS
// ============================================

// Si se ejecuta directamente (no como módulo)
if (require.main === module) {
    console.log('Resident Advisor API - Ejemplos de Código');
    console.log('=' .repeat(60));
    
    // Mostrar ejemplos bash
    mostrarEjemplosBash();
    
    // Ejecutar ejemplo JavaScript
    ejemploJavaScript().catch(console.error);
}

// Exportar para uso como módulo
module.exports = {
    ResidentAdvisorAPI,
    ejemplosBash,
    makeRequestWithRetry,
    processMultipleRequests,
    mostrarEjemplosBash,
    ejemploJavaScript
};

File: code/residentadvisor_api_examples.py
#!/usr/bin/env python3
"""
Ejemplos de código para interactuar con la API de Resident Advisor
Basado en la investigación de Stack Overflow
"""

import requests
import json
from typing import Dict, Any, Optional

class ResidentAdvisorAPI:
    """Cliente para la API de Resident Advisor GraphQL"""
    
    def __init__(self):
        self.endpoint = "https://ra.co/graphql"
        self.headers = {
            "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36",
            "Content-Type": "application/json"
        }
    
    def get_popular_events(self, filters: Optional[Dict] = None, page_size: int = 20) -> Dict[str, Any]:
        """
        Obtiene eventos populares usando GraphQL
        
        Args:
            filters: Diccionario con filtros (ej: {'area': 'London', 'listingDate': '2024-01-01'})
            page_size: Número de eventos a obtener
            
        Returns:
            Dict con la respuesta de la API
        """
        if filters is None:
            filters = {}
            
        query = """
        query GET_POPULAR_EVENTS($filters: FilterInputDtoInput, $pageSize: Int!) {
          popularEvents(filters: $filters, pageSize: $pageSize) {
            id
            title
            attending
            date
            contentUrl
            flyerFront
            images {
              id
              filename
              alt
              type
              crop
            }
            venue {
              id
              name
              contentUrl
              live
            }
          }
        }
        """
        
        variables = {
            "filters": filters,
            "pageSize": page_size
        }
        
        payload = {
            "query": query,
            "variables": variables
        }
        
        try:
            response = requests.post(
                self.endpoint,
                headers=self.headers,
                json=payload
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            print(f"Error al hacer request: {e}")
            return {}
    
    def get_dj_info(self, dj_id: str) -> Dict[str, Any]:
        """
        Obtiene información de un DJ específico (ejemplo genérico)
        Nota: Esta consulta requiere adaptación según los campos disponibles en GraphQL
        """
        query = """
        query GET_DJ_INFO($id: ID!) {
          dj(id: $id) {
            id
            name
            tracks {
              id
              title
              label {
                id
                name
              }
            }
            charts {
              id
              title
              position
            }
          }
        }
        """
        
        variables = {"id": dj_id}
        
        payload = {
            "query": query,
            "variables": variables
        }
        
        try:
            response = requests.post(
                self.endpoint,
                headers=self.headers,
                json=payload
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            print(f"Error al hacer request: {e}")
            return {}


def ejemplo_curl_completo():
    """Ejemplo de comando CURL completo para obtener eventos populares"""
    print("=== EJEMPLO COMANDO CURL COMPLETO ===")
    curl_command = """curl --location 'https://ra.co/graphql' \\
--header 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36' \\
--header 'Content-Type: application/json' \\
--data '{
  "query": "query GET_POPULAR_EVENTS($filters: FilterInputDtoInput, $pageSize: Int!) { popularEvents(filters: $filters, pageSize: $pageSize) { id title attending date contentUrl flyerFront images { id filename alt type crop } venue { id name contentUrl live } } }",
  "variables": {
    "filters": {},
    "pageSize": 20
  }
}'"""
    print(curl_command)
    print("\n")


def ejemplo_python_completo():
    """Ejemplo completo de uso en Python"""
    print("=== EJEMPLO PYTHON COMPLETO ===")
    
    # Inicializar cliente
    ra_api = ResidentAdvisorAPI()
    
    # Obtener eventos populares
    print("Obteniendo eventos populares...")
    eventos = ra_api.get_popular_events(page_size=10)
    
    if eventos and 'data' in eventos and 'popularEvents' in eventos['data']:
        print(f"Se encontraron {len(eventos['data']['popularEvents'])} eventos:")
        for evento in eventos['data']['popularEvents'][:3]:  # Solo los primeros 3
            print(f"- {evento['title']} ({evento['date']})")
            print(f"  Asistentes estimados: {evento['attending']}")
            print(f"  Lugar: {evento['venue']['name']}")
            print()
    else:
        print("No se pudieron obtener eventos o la respuesta no tiene la estructura esperada")


def ejemplo_filtros():
    """Ejemplo con filtros específicos"""
    print("=== EJEMPLO CON FILTROS ===")
    
    ra_api = ResidentAdvisorAPI()
    
    # Filtros por área
    filtros_london = {
        "area": "London"
    }
    
    print("Buscando eventos en London...")
    eventos_london = ra_api.get_popular_events(filters=filtros_london, page_size=5)
    
    if eventos_london and 'data' in eventos_london:
        print(f"Eventos en London: {len(eventos_london['data'].get('popularEvents', []))}")
    
    # Filtros por fecha de listado
    filtros_fecha = {
        "listingDate": "2024-11-01"
    }
    
    print("Buscando eventos listados después del 1 de noviembre...")
    eventos_fecha = ra_api.get_popular_events(filters=filtros_fecha, page_size=5)
    
    if eventos_fecha and 'data' in eventos_fecha:
        print(f"Eventos después de la fecha: {len(eventos_fecha['data'].get('popularEvents', []))}")


if __name__ == "__main__":
    print("=" * 60)
    print("EJEMPLOS DE USO - RESIDENT ADVISOR API")
    print("=" * 60)
    
    # Mostrar ejemplo CURL
    ejemplo_curl_completo()
    
    # Ejemplo Python básico
    print("Ejecutando ejemplo Python...")
    ejemplo_python_completo()
    
    print("\n" + "=" * 60)
    print("EJEMPLOS DE FILTROS")
    print("=" * 60)
    ejemplo_filtros()
    
    print("\n" + "=" * 60)
    print("NOTAS IMPORTANTES:")
    print("- Esta API no es oficial")
    print("- No requiere autenticación")
    print("- Usa user-agent de navegador legítimo")
    print("- Respeta límites de rate limiting")
    print("- La estructura puede cambiar sin aviso")
    print("=" * 60)